### PR TITLE
Add CaDiCaL support to CMake

### PR DESCRIPTION
**Pull request by tautschnig**
Add CMake configuration to build using CaDiCaL, which was already
supported with Makefiles. Also bump the version of CaDiCaL to sr19 and
fix the use of the cnf_solvert interface.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
FWIW: running the regression suite with CaDiCaL appears to be quite a bit slower, but an evaluation on more realistic benchmarks is yet to be done.

File: src/solvers/CMakeLists.txt [hunk 1]
```diff
@@ -98,6 +98,36 @@ elseif("${sat_impl}" STREQUAL "glucose")
     target_sources(solvers PRIVATE ${glucose_source})
 
     target_link_libraries(solvers glucose-condensed)
+elseif("${sat_impl}" STREQUAL "cadical")
+    message(STATUS "Building solvers with cadical")
+
+    download_project(PROJ cadical
+        URL https://github.com/arminbiere/cadical/archive/sr19.tar.gz
+        PATCH_COMMAND true
+        COMMAND CXX=${CXX_COMPILER} CXXFLAGS=-O3 ./configure --debug
```
Comment (smowton) [inline src/solvers/CMakeLists.txt]: Double-check: is the `--debug` asking `configure` for verbose output, or setting `-g` for the build, or could it even be overriding the `-O3` given in `CXXFLAGS` to get `-g -O0`?

**Review by smowton — APPROVED** [review]
Build script stuff looks fine to me, no opinion on the solvers stuff that I'm not that familiar with

File: src/solvers/CMakeLists.txt (hunk 1 above)
Comment (tautschnig) [reply src/solvers/CMakeLists.txt]: I did double-check when building locally: it just adds `-g`, the `-O3` from `CXXFLAGS` is passed on as desired.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7931e35fe6f23fd32dd6227231d3df26c7089f28).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114606145

**Review by kroening — APPROVED** [review]

File: src/solvers/sat/satcheck_cadical.cpp [hunk 2]
```diff
@@ -79,6 +79,21 @@ propt::resultt satcheck_cadicalt::do_prop_solve()
   }
   else
   {
+    // if assumptions contains false, we need this to be UNSAT
```
Comment (thk123) [inline src/solvers/sat/satcheck_cadical.cpp]: :pick: I don't understand this change or why the assume falses need special casing, but it should probably go in its own commit explaining in more detail why this is required

Comment (thk123) [reply src/solvers/sat/satcheck_cadical.cpp]: Preferably with a test that only passes with this

**Review by thk123 — COMMENTED** [review]
I don't have any strong opinions on this, but it should probably build `cadical` on CI and run the tests?

File: src/solvers/sat/satcheck_cadical.cpp [hunk 3]
```diff
@@ -120,13 +135,21 @@ satcheck_cadicalt::~satcheck_cadicalt()
...
 void satcheck_cadicalt::set_assumptions(const bvt &bv)
 {
-  INVARIANT(false, "method not supported");
+  assumptions = bv;
+
+  for(const auto &a : assumptions)
+  {
+    if(a.is_true())
+    {
+      assumptions.clear();
```
Comment (peterschrammel) [inline src/solvers/sat/satcheck_cadical.cpp]: :no_entry_sign: I don't understand this. Assumptions are connected conjunctively.

Comment (tautschnig) [reply src/solvers/sat/satcheck_cadical.cpp]: That was me copy&pasting from the `satcheck_minisat2.cpp` ... probably worth reviewing.

Comment (peterschrammel) [reply src/solvers/sat/satcheck_cadical.cpp]: See https://github.com/diffblue/cbmc/pull/4810

File: src/solvers/sat/satcheck_cadical.cpp (hunk 2 above)
Comment (tautschnig) [reply src/solvers/sat/satcheck_cadical.cpp]: Just for the record: this is done in keeping with other solvers, and thus about the way in which we use the solver interface. I agree that a test would be useful.

Comment (kroening) [reply src/solvers/sat/satcheck_cadical.cpp]: BTW, the reason why 'assume false' needs special casing is that SAT solvers don't have a 'false' literal.

**Review by martin-cs — APPROVED** [review]
Seems good to me.

**Review by peterschrammel — APPROVED** [review]
I'd suggest to merge also https://github.com/diffblue/cbmc/pull/5305 so that we can use the other solvers too without having to implement hardness_collectort everywhere.

File: src/solvers/CMakeLists.txt [hunk 4]
```diff
@@ -98,6 +98,36 @@ elseif("${sat_impl}" STREQUAL "glucose")
...
+
+    target_compile_definitions(solvers PUBLIC
+        SATCHECK_CADICAL HAVE_CADICAL
+    )
+
+    add_library(cadical STATIC IMPORTED)
+
+    set_target_properties(
+        cadical
+        PROPERTIES IMPORTED_LOCATION ${cadical_SOURCE_DIR}/build/libcadical.a
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/CMakeLists.txt:122]: This will only work on Linux, is that OK? If cadical also works on Windows/OSX it'd be nice if just used something like `find_library` here to get a more portable library path.

File: src/solvers/sat/satcheck_cadical.cpp [hunk 5]
```diff
@@ -23,7 +23,7 @@ tvt satcheck_cadicalt::l_get(literalt a) const
 
   tvt result;
 
-  if(a.var_no() > static_cast<unsigned>(solver->max()))
+  if(a.var_no() > static_cast<unsigned>(solver->vars()))
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/sat/satcheck_cadical.cpp]: Preferably use `narrow` from `util/narrow.h` here

File: src/solvers/sat/satcheck_cadical.cpp [hunk 6]
```diff
@@ -72,31 +89,37 @@ propt::resultt satcheck_cadicalt::do_prop_solve()
...
-          "solving inside CaDiCaL SAT solver has been interrupted");
+      log.status() << "got FALSE as assumption: instance is UNSATISFIABLE"
+                   << messaget::eom;
+      status = statust::UNSAT;
+      return resultt::P_UNSATISFIABLE;
     }
   }
 
+  for(const auto &a : assumptions)
+    solver->assume(a.dimacs());
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/sat/satcheck_cadical.cpp:106]: Why do we need to do this now? This wasn't here previously?

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
@tautschnig I'm noticing that the second commit has quite a few related changes at once. Is there any chance these could be split up into indivial commits that still make sense on their own? No real issue with the changes themselves.

File: src/solvers/sat/satcheck_cadical.cpp (hunk 6 above)
Comment (tautschnig) [reply src/solvers/sat/satcheck_cadical.cpp:106]: Support for assumptions simply was missing before.

File: src/solvers/sat/satcheck_cadical.cpp (hunk 5 above)
Comment (tautschnig) [reply src/solvers/sat/satcheck_cadical.cpp]: Will do.

**Comment by tautschnig** [comment]
> @tautschnig I'm noticing that the second commit has quite a few related changes at once. Is there any chance these could be split up into indivial commits that still make sense on their own? No real issue with the changes themselves.

I'm not entirely convinced this adds value: the previous implementation was pretty broken, and not really usable. All these changes are necessary to make things work.

File: src/solvers/CMakeLists.txt (hunk 4 above)
Comment (tautschnig) [reply src/solvers/CMakeLists.txt:122]: Cadical's makefile hard-codes this, so I don't think it'll work out of the box on other platforms.
